File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/ValidationRunStatusID.java
package net.nemerosa.ontrack.model.structure;

import lombok.Data;

import java.util.*;

@Data
public class ValidationRunStatusID {

    public static final String DEFECTIVE = "DEFECTIVE";
    public static final ValidationRunStatusID STATUS_DEFECTIVE = ValidationRunStatusID.of(DEFECTIVE, "Defective", false, false);
    public static final String EXPLAINED = "EXPLAINED";
    public static final ValidationRunStatusID STATUS_EXPLAINED = ValidationRunStatusID.of(EXPLAINED, "Explained", false, false);
    public static final String FAILED = "FAILED";
    public static final ValidationRunStatusID STATUS_FAILED = ValidationRunStatusID.of(FAILED, "Failed", true, false);
    public static final String FIXED = "FIXED";
    public static final ValidationRunStatusID STATUS_FIXED = ValidationRunStatusID.of(FIXED, "Fixed", false, true);
    public static final String INTERRUPTED = "INTERRUPTED";
    public static final ValidationRunStatusID STATUS_INTERRUPTED = ValidationRunStatusID.of(INTERRUPTED, "Interrupted", true, false);
    public static final String INVESTIGATING = "INVESTIGATING";
    public static final ValidationRunStatusID STATUS_INVESTIGATING = ValidationRunStatusID.of(INVESTIGATING, "Investigating", true, false);
    public static final String PASSED = "PASSED";
    public static final ValidationRunStatusID STATUS_PASSED = ValidationRunStatusID.of(PASSED, "Passed", true, true);
    public static final String WARNING = "WARNING";
    public static final ValidationRunStatusID STATUS_WARNING = ValidationRunStatusID.of(WARNING, "Warning", true, false);

    private final String id;
    private final String name;
    private final boolean root;
    private final boolean passed;
    private final Collection<String> followingStatuses;

    public ValidationRunStatusID addDependencies(String... followingStatuses) {
        List<String> dependencies = new ArrayList<>(this.followingStatuses);
        dependencies.addAll(Arrays.asList(followingStatuses));
        return new ValidationRunStatusID(id, name, root, passed, Collections.unmodifiableList(dependencies));
    }

    public static ValidationRunStatusID of(String id, String name, boolean root, boolean passed) {
        return new ValidationRunStatusID(id, name, root, passed, Collections.emptyList());
    }

}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/ValidationRunStatusID.kt
package net.nemerosa.ontrack.model.structure

import com.fasterxml.jackson.annotation.JsonProperty

class ValidationRunStatusID(
        val id: String,
        val name: String,
        @JsonProperty("root")
        val isRoot: Boolean,
        @JsonProperty("passed")
        val isPassed: Boolean,
        val followingStatuses: Collection<String> = emptyList()
) {

    companion object {
        const val DEFECTIVE = "DEFECTIVE"
        @JvmField
        val STATUS_DEFECTIVE = ValidationRunStatusID(DEFECTIVE, "Defective", isRoot = false, isPassed = false)
        const val EXPLAINED = "EXPLAINED"
        @JvmField
        val STATUS_EXPLAINED = ValidationRunStatusID(EXPLAINED, "Explained", false, isPassed = false)
        const val FAILED = "FAILED"
        @JvmField
        val STATUS_FAILED = ValidationRunStatusID(FAILED, "Failed", true, isPassed = false)
        const val FIXED = "FIXED"
        @JvmField
        val STATUS_FIXED = ValidationRunStatusID(FIXED, "Fixed", false, isPassed = true)
        const val INTERRUPTED = "INTERRUPTED"
        @JvmField
        val STATUS_INTERRUPTED = ValidationRunStatusID(INTERRUPTED, "Interrupted", isRoot = true, isPassed = false)
        const val INVESTIGATING = "INVESTIGATING"
        @JvmField
        val STATUS_INVESTIGATING = ValidationRunStatusID(INVESTIGATING, "Investigating", isRoot = true, isPassed = false)
        const val PASSED = "PASSED"
        @JvmField
        val STATUS_PASSED = ValidationRunStatusID(PASSED, "Passed", isRoot = true, isPassed = true)
        const val WARNING = "WARNING"
        @JvmField
        val STATUS_WARNING = ValidationRunStatusID(id = WARNING, name = "Warning", isRoot = true, isPassed = false)

        @JvmStatic
        fun of(id: String, name: String, root: Boolean, passed: Boolean): ValidationRunStatusID {
            return ValidationRunStatusID(id, name, root, passed, emptyList());
        }
    }

    fun addDependencies(vararg followingStatuses: String): ValidationRunStatusID {
        val dependencies = this.followingStatuses + followingStatuses
        return ValidationRunStatusID(id, name, isRoot, isPassed, dependencies)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ValidationRunStatusID) return false

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

}
